marconi-chain-index/src/Marconi/ChainIndex/Indexers.hs

```diff
@@ -638,94 +651,104 @@ mkIndexerStream' f coordinator =
-              whenNoIntersectionFound NoIntersectionFound = do
-                logError stdoutTrace $
-                  "No intersection found when looking for the chain point"
-                    <+> pretty resumePoint
+  :: [(Worker, Maybe FilePath)]
+  -> ChainIndexerT IO ()
+runIndexers indexerList = do
+  marconiTrace <- Lens.view runIndexerConfigTrace
+  retryConfig <- Lens.view runIndexerConfigRetryConfig
+  socketPath <- Lens.view runIndexerConfigSocketPath
+  indexingDepth <- Lens.view runIndexerConfigIndexingDepth
+  securityParam <- Lens.view runIndexerConfigSecurityParam
+  networkId <- Lens.view runIndexerConfigNetworkId
+  cliChainPoint <- Lens.view runIndexerConfigChainPoint
+  (ShouldFailIfResync shouldFailIfResync) <- Lens.view runIndexerConfigShouldFailIfResync
+  withNodeConnectRetryMarconi $ do
+    currentNodeBlockNo <-
+      liftIO
+        . Utils.toException
+        $ Utils.queryCurrentNodeBlockNo @Void networkId socketPath
+    let indexers = mapMaybe sequenceA indexerList
+    coordinator <- liftIO $ initializeCoordinatorFromIndexers securityParam indexingDepth indexers
+    let indexerDepth =
+          let SecurityParam s = securityParam
+           in case indexingDepth of
+                MinIndexingDepth d -> SecurityParam $ s - d + 1
+                MaxIndexingDepth -> SecurityParam 1
+    resumablePoints <-
+      liftIO $ getStartingPointsFromIndexers indexerDepth currentNodeBlockNo indexers coordinator
+    let oldestCommonChainPoint = minimum resumablePoints
+        resumePoint = case cliChainPoint of
+          C.ChainPointAtGenesis -> oldestCommonChainPoint -- User didn't specify a chain point, use oldest common chain point,
+          cliCp -> cliCp -- otherwise use what was provided on CLI.
+    logMInfo $
+      "Resumable points for each indexer:"
+        <> line
+        <> indent 4 (align (list (fmap pretty resumablePoints)))
+
+    -- Possible runtime failure if an indexer with a non-genesis resumable point will resume from
+    -- genesis.
+    if shouldFailIfResync
+      && elem C.ChainPointAtGenesis resumablePoints
+      && any (\case ChainPoint{} -> True; _ -> False) resumablePoints
+      then do
+        logMError $
+          nest
+            4
+            ( "At least one indexer has a non-genesis resumable point, while the oldest common resumable point between indexers is genesis."
+                <> line
+                <> "Are you sure you want to restart syncing that indexer from genesis?"
+                <> line
+                <> "If so, remove the '--fail-if-resyncing-from-genesis' flag."
+            )
+      else do
+        let stream =
+              mkIndexerStream coordinator
+                . chainSyncEventStreamLogging
+                . updateProcessedBlocksMetric
+            runChainSyncStream = liftIO $ withChainSyncBlockEventStream socketPath networkId [resumePoint] stream
```

Comment:
## Overview of one of the problems encountered in my investigation

`chainSyncEventStreamLogging` needs to have access to the logging context.

The `stream` is problematic here because it needs to both be in the `ChainIndexT` context (as noted above) and in `IO` because it is run by `withChainSyncBlockEventStream`. Inside, it uses `withAsync` on the stream consumer which means that we would need a way to "unlift" the `ChainIndexT` back to `IO`.

This brings forward the question of whether we can make `ChainIndexT` an instance of `MonadUnliftIO`. This would restrict the monad stack to allow only the current reader capability, and also it does not play well with the `ExceptT` we have decided to keep (as per @willjgould's investigation).

I believe that there are two possible ways to go on from here:
1. we make `ChainIndexT` an instance of `MonadUnliftIO` and see how it impacts the usage of `ExceptT`, keeping in mind the restrictions we will impose
2. we refactor the code in such a way that rids it of how deeply coupled it is with IO now, as Will also suggested; I don't know if this is possible, someone with more knowledge of the business logic should have a better idea than me